xds/src/main/java/io/grpc/xds/internal/sds/SslContextProviderSupplier.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.sds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.xds.EnvoyServerProtoData.UpstreamTlsContext;
+import io.netty.handler.ssl.SslContext;
+
+/**
+ * Enables the CDS policy to initialize this object with the received {@link UpstreamTlsContext} &
+ * communicate it to the consumer i.e. {@link SdsProtocolNegotiators.ClientSdsProtocolNegotiator}
+ * to lazily evaluate the {@link SslContextProvider}. The supplier prevents credentials leakage in
+ * cases where the user is not using xDS credentials but the CDS policy contains a non-default
+ * {@link UpstreamTlsContext}.
+ */
+public final class SslContextProviderSupplier implements Closeable {
+
+  private final UpstreamTlsContext upstreamTlsContext;
+  private final TlsContextManager tlsContextManager;
+  private SslContextProvider sslContextProvider;
+  private boolean shutdown;
+
+  public SslContextProviderSupplier(
+      UpstreamTlsContext upstreamTlsContext, TlsContextManager tlsContextManager) {
+    this.upstreamTlsContext = upstreamTlsContext;
+    this.tlsContextManager = tlsContextManager;
+  }
+
+  public UpstreamTlsContext getUpstreamTlsContext() {
+    return upstreamTlsContext;
+  }
+
+  @VisibleForTesting public boolean isShutdown() {
+    return shutdown;
+  }
+
+  /** Updates SslContext via the passed callback. */
+  public synchronized void updateSslContext(final SslContextProvider.Callback callback) {
+    checkNotNull(callback, "callback");
+    if (sslContextProvider == null && !shutdown) {
+      sslContextProvider =
+          tlsContextManager.findOrCreateClientSslContextProvider(upstreamTlsContext);
+    }
+    // we want to increment the ref-count so call findOrCreate again...
```

Comment:
So if `shutdown` is true, you still `findOrCreate()` and `addCallback()`?

Reply:
Yes. Ideally this should not happen because after `shutdown` is true no CDS load balancer should have any reference to the Supplier and so there should be no call to `updateSslContext`. I guess I can throw an `IllegalStateException` but it's doesn't seem that important to do anything like that. WDYT?

Reply:
done